common/geometry/gpu_distance.py

```diff
@@ -0,0 +1,264 @@
+    num_groups: int,
+) -> torch.Tensor:
+    """Compute minimum distance from each point to each group of segments.
+
+    This is useful for computing distance to polylines (groups of connected
+    segments) or polygon boundaries.
+
+    Args:
+        query_points: (Q, 2) tensor of query point coordinates
+        segment_starts: (S, 2) tensor of segment start coordinates
+        segment_ends: (S, 2) tensor of segment end coordinates
+        segment_to_group: (S,) tensor mapping each segment to its group index
+        num_groups: Number of groups
+
+    Returns:
+        (Q, num_groups) tensor where result[i, j] is the minimum distance
+        from query_points[i] to any segment in group j.
+    """
+    # Compute all pairwise distances, shape (Q, S)
+    all_distances = point_to_segment_distance(query_points, segment_starts, segment_ends)
+
+    Q = query_points.shape[0]
+    device = query_points.device
+
+    # Initialize with infinity
+    min_distances = torch.full(
+        (Q, num_groups), float("inf"), device=device, dtype=all_distances.dtype
+    )
+
+    # Expand segment_to_group to match all_distances shape for scatter
+    # segment_to_group: (S,) -> (Q, S)
+    group_indices = segment_to_group.unsqueeze(0).expand(Q, -1)
+
+    # Scatter minimum distances into groups
+    min_distances.scatter_reduce_(1, group_indices, all_distances, reduce="amin")
+
+    return min_distances
+
+
+def point_in_polygon_winding(
+    query_points: torch.Tensor,
+    polygon_vertices: torch.Tensor,
+    polygon_ranges: torch.Tensor,
+) -> torch.Tensor:
+    """Test if points are inside polygons using winding number algorithm.
+
+    The winding number counts how many times the polygon winds around the point.
+    A non-zero winding number means the point is inside.
+
+    Args:
+        query_points: (Q, 2) tensor of query point coordinates
+        polygon_vertices: (V, 2) tensor of all polygon vertices concatenated
+        polygon_ranges: (P, 2) tensor where each row is [start, end) indices
+            into polygon_vertices for that polygon's vertices
+
+    Returns:
+        (Q, P) boolean tensor where result[i, j] is True if query_points[i]
+        is inside polygon j.
+    """
+    Q = query_points.shape[0]
+    P = polygon_ranges.shape[0]
+    device = query_points.device
+
+    # Initialize winding numbers
+    winding = torch.zeros((Q, P), device=device, dtype=torch.float32)
+
+    # Process each polygon
+    for poly_idx in range(P):
+        start_idx = polygon_ranges[poly_idx, 0].item()
+        end_idx = polygon_ranges[poly_idx, 1].item()
+
+        if end_idx <= start_idx:
+            continue
+
+        # Get polygon vertices for this polygon
+        poly_verts = polygon_vertices[start_idx:end_idx]  # (N, 2)
+        N = poly_verts.shape[0]
+
+        # Create edges: from vertex i to vertex (i+1) % N
+        v1 = poly_verts  # (N, 2)
+        v2 = torch.roll(poly_verts, -1, dims=0)  # (N, 2)
```

Comment:
I see, I believe this means that the first and last point of the segment are assumed to be connected.